zebra-chain/src/commitments/sapling.rs

```diff
@@ -0,0 +1,341 @@
+    const J: [u8; 8] = *b"Zcash_J_";
+
+    P + find_group_hash(J, b"") * x
+}
+
+/// Construct a 'windowed' Pedersen commitment by reusing a Pederson hash
+/// construction, and adding a randomized point on the Jubjub curve.
+///
+/// WindowedPedersenCommit_r (s) := \
+///   PedersenHashToPoint(“Zcash_PH”, s) + [r]FindGroupHash^J^(r)(“Zcash_PH”, “r”)
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretewindowedcommit
+pub fn windowed_pedersen_commitment(r: jubjub::Fr, s: &BitVec<Lsb0, u8>) -> jubjub::ExtendedPoint {
+    const D: [u8; 8] = *b"Zcash_PH";
+
+    pedersen_hash_to_point(D, &s) + find_group_hash(D, b"r") * r
+}
+
+/// The randomness used in the Pedersen Hash for note commitment.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct CommitmentRandomness(jubjub::Fr);
+
+/// Note commitments for the output notes.
+#[derive(Clone, Copy, Deserialize, PartialEq, Serialize)]
+pub struct NoteCommitment(#[serde(with = "serde_helpers::AffinePoint")] pub jubjub::AffinePoint);
+
+impl fmt::Debug for NoteCommitment {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("NoteCommitment")
+            .field("u", &hex::encode(self.0.get_u().to_bytes()))
+            .field("v", &hex::encode(self.0.get_v().to_bytes()))
+            .finish()
+    }
+}
+
+impl From<[u8; 32]> for NoteCommitment {
+    fn from(bytes: [u8; 32]) -> Self {
+        Self(jubjub::AffinePoint::from_bytes(bytes).unwrap())
+    }
+}
+
+impl From<jubjub::ExtendedPoint> for NoteCommitment {
+    fn from(extended_point: jubjub::ExtendedPoint) -> Self {
+        Self(jubjub::AffinePoint::from(extended_point))
+    }
+}
+
+impl From<NoteCommitment> for [u8; 32] {
+    fn from(cm: NoteCommitment) -> [u8; 32] {
+        cm.0.to_bytes()
+    }
+}
+
+impl Eq for NoteCommitment {}
+
+impl ZcashSerialize for NoteCommitment {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_all(&self.0.to_bytes())?;
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for NoteCommitment {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        Ok(Self(
+            jubjub::AffinePoint::from_bytes(reader.read_32_bytes()?).unwrap(),
+        ))
+    }
+}
+
+impl NoteCommitment {
+    /// Generate a new _NoteCommitment_ and the randomness used to create it.
+    ///
+    /// We return the randomness because it is needed to construct a _Note_,
+    /// before it is encrypted as part of an _Output Description_.
+    ///
+    /// NoteCommit^Sapling_rcm (g*_d , pk*_d , v) :=
+    ///   WindowedPedersenCommit_rcm([1; 6] || I2LEBSP_64(v) || g*_d || pk*_d)
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#concretewindowedcommit
+    #[allow(non_snake_case)]
+    pub fn new<T>(
+        csprng: &mut T,
+        diversifier: Diversifier,
+        transmission_key: TransmissionKey,
+        value: Amount<NonNegative>,
```

Comment:
Ah yes, we do this here! And this is supposed to be the same value input into `ValueCommitment`

Reply:
It's turned into bytes here, not a scalar, though

Reply:
What's this? I think I'm missing an xref from another conversation.

Reply:
Ah sorry, the discussion above about `TODO: accept an Amount instead of just bytes?` re: constructing a `NoteCommitment`